### PR TITLE
Significant general performance optimizations in some cases

### LINKED ISSUE #584: Question: Why not `{ RETURN_DOM_FRAGMENT: true, RETURN_DOM_IMPORT: false }`?
The readme states there are XSS vulnerabilities when `RETURN_DOM_FRAGMENT` is `true` while `RETURN_DOM_IMPORT` is `false`. Out of curiosity, could you give any example(s) of this? And an explanation of why it occurs or why `RETURN_DOM_IMPORT` prevents it would be appreciated :)

### LINKED ISSUE #585: Questions About `_isClobbered`
Out of curiosity, could you give an example of some input that makes the condition at https://github.com/cure53/DOMPurify/blob/dbf7a34152b9502ef0b828f194dcdd8bd54282a0/src/purify.js#L793 return `true`?

Also, why does this still run when the `SANITIZE_DOM` config option is `false`? I was under the impression that setting `SANITIZE_DOM` to `false` disables any clobbering-related checks. Is that not the purpose of `SANITIZE_DOM`? If not, what's the purpose?

### PR DESCRIPTION
**Pull request by GrantGryczan**
Resolves #584
Resolves #585 

Incidentally, this also ensures `importNode` is only called on the output when necessary, rather than letting users configure it with `RETURN_DOM_IMPORT`, possibly opening themselves up to vulnerabilities. (See discussion below.)

### CODE REVIEW

**Review by GrantGryczan — COMMENTED** [review]
I need your confirmation on the last commit:

I think it would be better to have it always call `importNode` if `RETURN_DOM_IMPORT` is set to `true`, so that it works as expected when set explicitly, but have it _default_ to `false` when `shadowroot` is not in the `ALLOWED_ATTR` attribute, rather than overriding it and preventing `importNode` from getting called even when `RETURN_DOM_IMPORT` is `true`. Does that sound appropriate?

If so, some tests will need to be changed in consideration of this change, but I can probably do that.

**Comment by cure53** [comment]
Alright, let's run some tests :)

**Comment by GrantGryczan** [comment]
It's not ready yet; see my last comment.

**Comment by cure53** [comment]
Oh, sorry, no worries.

> I think it would be better to have it always call importNode if RETURN_DOM_IMPORT is set to true, so that it works as expected when set explicitly, but have it default to false when shadowroot is not in the ALLOWED_ATTR attribute, rather than overriding it and preventing importNode from getting called even when RETURN_DOM_IMPORT is true. Does that sound appropriate?

So, the goal of whatever we do here is to avoid having people configure themselves to be prone to XSS when they are not aware of it. No matter how we do it, this needs to be prio number one.

**Comment by GrantGryczan** [comment]
> So, the goal of whatever we do here is to avoid having people configure themselves to be prone to XSS when they are not aware of it. No matter how we do it, this needs to be prio number one.

In that case, perhaps it would be better to discard the `RETURN_DOM_IMPORT` option entirely, in favor of forcing it to `importNode` when the `shadowroot` attribute is allowed?

**Comment by cure53** [comment]
Hmmm, not sure how many people already use it. And not sure how many of those allow `shadowroot` :D

**Comment by GrantGryczan** [comment]
Of the people who are using it, there are two possible cases:
1. They allow `shadowroot`. Then removing the option fixes an XSS vulnerability for them.
2. They don't allow `shadowroot`. Then removing the option changes nothing and the config option is a no-op for them.

I don't see any issue with removing it.

**Comment by cure53** [comment]
I think you are right, let's nuke it.

**Comment by GrantGryczan** [comment]
Oh! I realize the reason it's there despite your concern about users unwittingly opening themselves to vulnerability. It's useful if all of the following conditions are true for your use case:

1. You have `RETURN_DOM` or `RETURN_DOM_FRAGMENT` enabled (obviously).
2. You allow the `shadowroot` attribute.
3. You are NOT appending the DOM node outputted by DOMPurify into any document. (This one is the one we forgot about.)
4. You want to improve performance by disabling the `importNode` call.

I still don't think it's worth keeping though, because it's a minor benefit (performance improvements for very large node trees) and only in very rare conditions (`shadowroot` allowed, not appending to any document). But it's up to you; let me know what you think.

**Comment by cure53** [comment]
Let's remove it for now and see if people complain. Less complexity is better in the long run.

**Comment by GrantGryczan** [comment]
Alright, all ready now. I've changed up a lot of the tests which used to pertain to DOM importing. Needless to say, review my changes to the tests especially carefully.

File: README.md [hunk 1]
```diff
@@ -342,7 +332,7 @@ Many people helped and help DOMPurify become what it is and need to be acknowled
...
 
-## Testing powered by
```
Comment (GrantGryczan) [inline README.md:345]: Haha, looks like I unwittingly removed a couple instances of trailing whitespace. I have a shortcut configured in my editor to remove trailing whitespace automatically on save.

**Comment by GrantGryczan** [comment]
Looks like there are merge conflicts now that you've accepted a PR from someone else in the meantime. I hope it's easier for you to take care of that.

**Comment by cure53** [comment]
There should be nothing that conflicts with your changes afaics, just generated stuff

**Comment by cure53** [comment]
Alright, conflicts resolved, let's see what the tests say!

**Comment by cure53** [comment]
This looks great, thank you very much :bow: :smile:

**Pull request merged**